Account for pending request state in <img>.complete

twilco opened the pull request:
Per spec update, if the \<img\> current request is completely available or broken, `complete` is only `true` if the \<img\>'s pending request is also unavailable.

> The IDL attribute complete must return true if any of the following conditions is true:
       * Both the src attribute and the srcset attribute are omitted.
       * The srcset attribute is omitted and the src attribute's value is the empty string. 
       * **The img element's current request's state is completely available and its pending request is null.**
       * **The img element's current request's state is broken and its pending request is null.** 

Spec diff: https://whatpr.org/html/4934/1ab7c86...02772ea/embedded-content.html#the-img-element

#24450

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #24450.
- [x] These changes do not require tests because they _should_ be covered by the automatically synced WPT associated with the spec update at https://github.com/whatwg/html/pull/4934.  

However, the [test updated with the spec change](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html) passes before and after this PR.  Should another test be added?

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @paulrouget (or someone else) soon.

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

jdm commented:
So the test for the pending request in particular did not actually change in the WPT PR: https://github.com/web-platform-tests/wpt/pull/19298/files#diff-87cc01146c1ed045cec3a26399e01aadL197-R200

The fact that we already passed the "async pending request test" before these changes is a bit unnerving. It suggests that our code that deals with setting `src` to a new value already modifies the current request, which would be incorrect.

twilco commented:
Hey @jdm — thanks for the follow up.

> So the test for the pending request in particular did not actually change in the WPT PR

Ah, oops.  Is this test sufficient for this PR, or is more needed?

> The fact that we already passed the "async pending request test" before these changes is a bit unnerving. It suggests that our code that deals with setting src to a new value already modifies the current request, which would be incorrect.

Just to confirm, I tried again off the latest `master` (b5759e711348dfa1dab86d6e60940e2db720d259), and this test indeed passes.  Is this something we should track down here, or should another issue be created?

Also, the [Appveyor build](https://ci.appveyor.com/project/servo/servo/builds/28159481) appears to have failed due to timeout:

> Build execution time has reached the maximum allowed time for your plan (100 minutes).

Does this require any action on my part besides a retry?  If not, how do I issue a retry?

jdm commented:
Appveyor usually times out these days. We should probably just disable it.

jdm commented:
Oh, it took me about 15 seconds after looking at htmlimageelement.rs to figure out why we passed that test already:
```
        // https://html.spec.whatwg.org/multipage/#reacting-to-dom-mutations
        // Always first set the current request to unavailable,
        // ensuring img.complete is false.
        {
            let mut current_request = self.current_request.borrow_mut();
            current_request.state = State::Unavailable;
        }
```
That code originated in https://github.com/servo/servo/commit/71b0c1016461ab51daa9fc13e6454be1c8553c97. Reading https://html.spec.whatwg.org/multipage/images.html#when-to-obtain-images and https://html.spec.whatwg.org/multipage/images.html#update-the-image-data I'm having trouble figuring out why the test expects that img.complete should return false immediately after setting the image's src attribute.

Since that test came from https://github.com/web-platform-tests/wpt/commit/95283d44677b8be60680bc00859abca8ec83c85e which came from https://bugzilla.mozilla.org/show_bug.cgi?id=1578419. I'm going to follow up there to see if Boris can point out something that I'm missing.

jdm commented:
I propose we update that code that I pointed out to change the pending request, rather than the current one. Then we can keep an eye on https://github.com/whatwg/html/issues/4884 and ensure that our implementation matches the eventual specification changes.

twilco commented:
Thanks for investigating this one!

With the pre-existing commit on this PR, changing this:

```
// https://html.spec.whatwg.org/multipage/#reacting-to-dom-mutations
// Always first set the current request to unavailable,
// ensuring img.complete is false.
{
    let mut current_request = self.current_request.borrow_mut();
    current_request.state = State::Unavailable;
}
```

to:

```
{
    let mut pending_request = self.pending_request.borrow_mut();
    pending_request.state = State::Unavailable;
}
```

Results in the following test failures, all of which used to pass:

```
$ ./mach test-wpt --headless tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element
Running 45 tests in web-platform-tests

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async pending request test
  │   → assert_false: Should not be complete because we have started a new load expected false got true
  │
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:197:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-media.html:
  │ FAIL [expected PASS] set media after src updates selected image
  │   → assert_true: expected true got false
  │
  │ onImgLoad@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-media.html:22:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-src-complete.html:
  │ FAIL [expected PASS] Changing the img src should retain the 'complete' property
  │   → uncaught exception: unknown (can't convert to string)
  │
  │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:12:9
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:20:9

Ran 45 tests finished in 100.0 seconds.
  • 42 ran as expected. 0 tests skipped.
  • 3 tests had unexpected subtest results
```

In order, these tests are:

https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html#L191#L199
https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/update-media.html
https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/update-src-complete.html

The first failure makes sense based on points you've raised in previous comments, namely:

> I'm having trouble figuring out why the test expects that img.complete should return false immediately after setting the image's src attribute.

The other two I'm less certain on — hoping to dig into them more in conjunction with the assorted specs.

jdm commented:
In theory the change you made should allow the test to continue passing, because this PR causes img.complete to check the status of the pending request as well. The fact that it's failing is very surprising!

https://github.com/web-platform-tests/wpt/blob/master/html/semantics/embedded-content/the-img-element/update-src-complete.html looks like it's failing for the exact same reason, since it's doing a very similar test.

twilco commented:
Indeed!  I'm at Rust Belt Rust for the next couple days but will dig in soon.

twilco commented:
Hey @jdm, struggling with this one a bit.

These tests are failing because we used to reset the `current_request.state` to `Unavailable` in the beginning of [`update_the_image_data`](https://github.com/servo/servo/blob/927dfd15bf87566f860a2ab6f4fcb507e83f8b1a/components/script/dom/htmlimageelement.rs#L890), but now we reset `pending_request` to this state.  Then, when calling `<img>.complete` the second time, `current_request` is still in a `CompletelyAvailable` state, and `pending_request` is `Unavailable`, resulting in a `true` evaluation for these back-to-back loads.  The change currently in this PR unfortunately doesn't prevent this from happening.

I've also tried changing this in the [`Complete`](https://github.com/servo/servo/blob/master/components/script/dom/htmlimageelement.rs#L1546#L1551) function:

```rust
        let image_request = self.image_request.get();
        match cur_request.state {
            State::CompletelyAvailable | State::Broken => {
                return image_request != ImageRequestPhase::Pending;
            },
            State::PartiallyAvailable | State::Unavailable => return false,
        }
```

My thinking was that an `ImageRequestPhase::Complete` would be equivalent to what the spec calls pending request being in a state of "null"...but this was mostly an unfounded guess based on other code I've seen in this file 🙂 .  It did not fix the problem — `image_request` was `ImageRequestPhase::Current` in both the initial load and the follow up load.

Some questions:

  **1.** The spec change often mentions the state of being `null` in regards to image requests (for example, in the spec diff).  Is `null` the same as the `State::Unavailable` state in Servo?  I've made that assumption, but I'm not sure that it's correct.  As mentioned before, I also experimented with `ImageRequestPhase` representing this null state.

  **2.** Is there any guidance you could provide on how to progress?  I've traced the codepath starting from [`attribute_mutated`](https://github.com/servo/servo/blob/927dfd15bf87566f860a2ab6f4fcb507e83f8b1a/components/script/dom/htmlimageelement.rs#L1617) to the final step of updating image data in conjunction with the spec, but I'm not sure what needs changing to account for the fact that we no longer reset `current_request.state` to `Unavailable` in the very beginning of the process.

twilco committed:
Properly represent null-state of pending request

The <img> spec frequently mentions setting and checking the
null-status of pending request, but prior to this commit we did not have
a consistent way to represent this.  This commit changes the
`pending_request` to be an DomRefCell<Option<ImageRequest>>, where a None value corresponds to a
null pending request in the specification.

<img>.complete is now only true if pending request is null (None).

Issue #24450

twilco commented:
Hey again @jdm!

I've pushed a commit that changes `pending_request` to a `DomRefCell<Option<ImageRequest>>`.   This unfortunately does not get us any further as far as passing tests go, with the same three failing:

```
$ ./mach test-wpt --headless tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element
Running 45 tests in web-platform-tests

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async pending request test
  │   → assert_false: Should not be complete because we have started a new load expected false got true
  │
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:197:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-media.html:
  │ FAIL [expected PASS] set media after src updates selected image
  │   → assert_true: expected true got false
  │
  │ onImgLoad@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-media.html:22:9
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-src-complete.html:
  │ FAIL [expected PASS] Changing the img src should retain the 'complete' property
  │   → uncaught exception: unknown (can't convert to string)
  │
  │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:12:9
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:20:9

Ran 45 tests finished in 100.0 seconds.
  • 42 ran as expected. 0 tests skipped.
  • 3 tests had unexpected subtest results
```

Note I've removed the initialization of `current_request` to `State::Unavailable` at the beginning of `update_the_image_data` per the suggestion in an above comment.  Re-adding this initialization, along with the changes currently in this PR, results in consistent success of all img WPT tests.

I tried adding some `dbg!`s and running just [this test](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/img.complete.html#L191#L199), and it looks like the second `.complete` call is executed before we've reached a stable state in preparation for step 8 of [updating the image data](https://html.spec.whatwg.org/multipage/#updating-the-image-data).  As far as I can tell, there isn't any state changes by this point that should result in `.complete` returning false, so without the `current_request.state = State::Unavailable` in the beginning we can't pass these tests.   [This comment](https://github.com/whatwg/html/issues/4884#issuecomment-544942138) in the issue you linked previously _seems_ to imply the same thing.  But maybe I'm missing something in the spec.

What are your thoughts?

twilco commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -849,8 +905,12 @@ impl HTMLImageElement {
                         // request to the broken state is done prior to queuing a
                         // task, why is this here?
+                        // Moving this outside the task results in intermittent failure of
+                        // non-active-document.html tests.
```
Not sure that this if this helpful or just noise — would be open to removing it and the other similar comment.

jdm replied:
Please file an issue about this with any additional information you have. This sounds like something that's worth investigating.

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -866,14 +926,6 @@ impl HTMLImageElement {
         let src = elem.get_url_attribute(&local_name!("src"));
         let base_url = document.base_url();
 
-        // https://html.spec.whatwg.org/multipage/#reacting-to-dom-mutations
```
I propose that we resurrect this block, but we change pending_request to Unavailable instead.

twilco replied:
Based on the latest WPT run, looks like we're back to to the two `<img>.complete` failures.  I'm guessing this suggestion aimed to resolve those failures, so I gave it a go.

---

**Attempt one:**
_Only setting `State::Unavailable` if `pending_request` is `Some`:_
```rust
{
    if let Some(ref mut pending_request) = *self.pending_request.borrow_mut() {
        pending_request.state = State::Unavailable;
    }
}
```

**Results:**
_Excludes the `assert_unreached: ref currentSrc was "" Reached unreachable code` you resolved separately_
```
Ran 45 tests finished in 101.0 seconds.
  • 38 ran as expected. 0 tests skipped.
  • 7 tests had unexpected subtest results

Tests with unexpected results:
  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async pending request test
  │   → assert_false: Should not be complete because we have started a new load expected false got true
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:197:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-src-complete.html:
  │ FAIL [expected PASS] Changing the img src should retain the 'complete' property
  │   → uncaught exception: unknown (can't convert to string)
  │ 
  │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:12:9
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:20:9

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: source (max-width:500px) broken image, img valid image, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: source (max-width:500px) valid image, img broken image, resize to narrow

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to narrow
```

---

**Attempt two:**
_Unconditionally setting `pending_request` to `State::Unavailable`, initializing `pending_request` to `ImageRequest::new()` if it is `None`:_

```rust
{
    let mut pending_opt = self.pending_request.borrow_mut();
    pending_opt.get_or_insert(ImageRequest::new()).state = State::Unavailable;
}
```

**Results:**

```
Running 45 tests in web-platform-tests

  /html/semantics/embedded-content/the-img-element/img-with-containment-and-size.html
  /html/semantics/embedded-content/the-img-element/document-base-url.html
  /html/semantics/embedded-content/the-img-element/document-adopt-base-url.html
  /html/semantics/embedded-content/the-img-element/available-images.html
  /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-002.html
  /html/semantics/embedded-content/the-img-element/sizes/sizes-dynamic-001.html
  /html/semantics/embedded-content/the-img-element/404-response-with-actual-image-data.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-path-changes.html
  /html/semantics/embedded-content/the-img-element/Image-constructor.html
  /html/semantics/embedded-content/the-img-element/adoption.html
  /html/semantics/embedded-content/the-img-element/current-pixel-density/error.html
  /html/semantics/embedded-content/the-img-element/current-pixel-density/basic.html
  /html/semantics/embedded-content/the-img-element/null-image-source.html
  /html/semantics/embedded-content/the-img-element/data-url.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-picture.html
  /html/semantics/embedded-content/the-img-element/historical-progress-event.window.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-svg.tentative.html
  /html/semantics/embedded-content/the-img-element/available-images-onload.html
  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async srcset complete test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:108:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src removal test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:161:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src broken test
  │   → assert_true: expected true got false
  │ 
  │ img.onerror<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:151:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async pending request test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:195:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async srcset removal test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:173:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src complete test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:89:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  /html/semantics/embedded-content/the-img-element/decode/image-decode.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-with-quick-attach.html
  /html/semantics/embedded-content/the-img-element/invalid-src.html
  /html/semantics/embedded-content/the-img-element/nonexistent-image.html
  /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-quirks-mode.html
  /html/semantics/embedded-content/the-img-element/delay-load-event-detached.html
  /html/semantics/embedded-content/the-img-element/relevant-mutations.html
  /html/semantics/embedded-content/the-img-element/delay-load-event.html
  /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-width-1000px.html
  /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-display-none.html
  /html/semantics/embedded-content/the-img-element/srcset/select-an-image-source.html
  /html/semantics/embedded-content/the-img-element/srcset/parse-a-srcset-attribute.html
  /html/semantics/embedded-content/the-img-element/non-active-document.html
  /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html
  /html/semantics/embedded-content/the-img-element/srcset/srcset-media-dynamic.html
  /html/semantics/embedded-content/the-img-element/usemap-casing.html
  /html/semantics/embedded-content/the-img-element/srcset/avoid-reload-on-resize.html
  /html/semantics/embedded-content/the-img-element/not-rendered-dimension-getter.html
  /html/semantics/embedded-content/the-img-element/update-the-image-data/fail-to-resolve.html
  /html/semantics/embedded-content/the-img-element/update-media.html
  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-src-complete.html:
  │ FAIL [expected PASS] Changing the img src should retain the 'complete' property
  │   → uncaught exception: unknown (can't convert to string)
  │ 
  │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:10:9
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:20:9

  /html/semantics/embedded-content/the-img-element/update-the-source-set.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-iframe.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-path-changes-svg.tentative.html
  /html/semantics/embedded-content/the-img-element/decode/image-decode-with-quick-attach-svg.tentative.html
  ▶ TIMEOUT [expected OK] /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html
  │ 
  │ 
  │ 
  └ 

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] img (srcset 1 cand) valid image, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) broken image, img valid image, resize to wide
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to wide
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: same URL in source (max-width:500px) and img, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] img (srcset 1 cand) valid image, resize to narrow

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img broken image, resize to narrow
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to narrow
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: same URL in source (max-width:500px) and img, resize to narrow


Ran 45 tests finished in 103.0 seconds.
  • 42 ran as expected. 0 tests skipped.
  • 1 tests timed out unexpectedly
  • 3 tests had unexpected subtest results

Tests with unexpected results:
  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async srcset complete test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:108:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src removal test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:161:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src broken test
  │   → assert_true: expected true got false
  │ 
  │ img.onerror<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:151:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async pending request test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:195:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async srcset removal test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:173:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/img.complete.html:
  │ FAIL [expected PASS] async src complete test
  │   → assert_true: expected true got false
  │ 
  │ img.onload<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/img.complete.html:89:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:1948:32

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/update-src-complete.html:
  │ FAIL [expected PASS] Changing the img src should retain the 'complete' property
  │   → uncaught exception: unknown (can't convert to string)
  │ 
  │ check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:10:9
  └ onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/update-src-complete.html:20:9

  ▶ TIMEOUT [expected OK] /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html
  │ 
  │ 
  │ 
  └ 

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] img (srcset 1 cand) valid image, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) broken image, img valid image, resize to wide
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to wide
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: same URL in source (max-width:500px) and img, resize to wide

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] img (srcset 1 cand) valid image, resize to narrow

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img broken image, resize to narrow
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  │ TIMEOUT [expected FAIL] picture: source (max-width:500px) valid image, img valid image, resize to narrow
  └   → Test timed out

  ▶ Unexpected subtest result in /html/semantics/embedded-content/the-img-element/environment-changes/viewport-change.html:
  └ PASS [expected FAIL] picture: same URL in source (max-width:500px) and img, resize to narrow
```

These results are worse, failing on the first `<img>.complete` in tests are where there are two `<img>.complete` calls.

---

Are either of these what you were thinking?  I've pushed the code for attempt one in case that helps illustrate the situation.  I can try to dig deeper later to see if there's some way we can make this work and figure out why these code changes have these results, but wanted to update with my initial findings.

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -1132,26 +1185,26 @@ impl HTMLImageElement {
                 if relevant_mutation {
-                    this.abort_request(State::Unavailable, ImageRequestPhase::Pending);
+                    let mut pending_opt = this.pending_request.borrow_mut();
+                    *pending_opt = None;
```
We could write this (and other similar uses) as `*this.pending_request.borrow_mut() = None;`. Up to you if you think that's an improvement.

twilco replied:
That's definitely better.  Will change.

jdm reviewed (changes requested):
Thanks! I know there's a lot of code here and the spec is a bit confusing at times. This looks like a sensible way to go about making this change.

twilco commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -742,39 +782,51 @@ impl HTMLImageElement {
+                    // Step 12
+                    if let Some(pending_url) = request.parsed_url.clone() {
+                        if pending_url == *url {
+                            return;
```
Good catch, will fix.

twilco commented:
I've made the stylistic change you suggested.  The rest of the changes cause tests to start failing again (😢), so I'll have to work through them and maybe walk back some of the other changes in this PR in an attempt to fix them.

twilco commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -366,6 +383,17 @@ impl HTMLImageElement {
+            &mut self.current_request.borrow_mut().deref_mut(),
+            &mut pending_opt.get_or_insert(ImageRequest::new()),
+        );
+        *pending_opt = None;
```
I like this suggestion a lot — definitely cleaner, and it is probably a good idea to bail rather than set current request to `ImageRequest::new()` when pending request is `None`.

However, this change has somehow resulted in a regression in some tests.  The behavior _is_ different, but (as far as I can tell) only when `pending_request` is `None`.  I added a `dbg!` to check for `None` in one of the failing tests to see if this was happening in either case.

**Before your suggestion:**
```rust
fn upgrade_pending_to_current(&self) {
    let mut pending_opt = self.pending_request.borrow_mut();
    dbg!(&pending_opt.is_none());
    mem::swap(
        &mut self.current_request.borrow_mut().deref_mut(),
        &mut pending_opt.get_or_insert(ImageRequest::new()),
    );
    *pending_opt = None;
    self.image_request.set(ImageRequestPhase::Current);
}
```

**After your suggestion:**
```rust
fn upgrade_pending_to_current(&self) {
    dbg!(&self.pending_request.borrow().is_none());
    *self.current_request.borrow_mut() = self.pending_request.borrow_mut().take().expect("no pending request");
    self.image_request.set(ImageRequestPhase::Current);
}
```

I ran against [just the parse-a-sizes-attribute-standards-mode.html test](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html), although a few other similar tests failed with similar errors.  I used this command:

```sh
./mach test-wpt --headless /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html
```

In both tests, the `dbg!` output showed `is_none()` was never true and printed the same number of lines, meaning the behavior before and after this change _should_ be the same.  As expected, there were never any crashes logged from the `.expect`, either.

**Output before your suggestion:**

```
0:04.24 INFO Using 8 client processes
 0:04.42 INFO Starting http server on 127.0.0.1:8000
 0:04.42 INFO Starting http server on 127.0.0.1:8001
 0:04.54 INFO Starting https server on 127.0.0.1:8443
 0:05.91 SUITE_START: web-platform-test - running 1 tests
 0:05.92 INFO Starting runner
 0:05.93 TEST_START: /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html
 0:05.93 WARNING Got command init_succeeded in state running
 0:06.77 pid:10971 Full command: /Users/twilco/projects/servo/target/debug/servo --hard-fail -u Servo/wptrunner -Z replace-surrogates -z http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html --user-stylesheet /Users/twilco/projects/servo/resources/ahem.css --certificate-path /Users/twilco/projects/servo/tests/wpt/web-platform-tests/tools/certs/cacert.pem
pid:10971 
 0:06.77 pid:10971 
 0:06.77 pid:10971 
 0:10.07 pid:10971 [components/script/dom/htmlimageelement.rs:398] &pending_opt.is_none() = false
 0:10.07 pid:10971 [components/script/dom/htmlimageelement.rs:398] &pending_opt.is_none() = false
 0:10.07 pid:10971 [components/script/dom/htmlimageelement.rs:398] &pending_opt.is_none() = false
 0:10.07 pid:10971 [components/script/dom/htmlimageelement.rs:398] &pending_opt.is_none() = false
...lots more false dbg statements...
 0:11.14 TEST_END: Test OK. Subtests passed 154/169. Unexpected 0
 0:11.14 INFO No more tests
 0:11.15 INFO Closing logging queue
 0:11.15 INFO queue closed
 0:11.15 INFO Got 0 unexpected results
 0:11.15 SUITE_END
```

**Output after your suggestion:**

```
 0:01.86 INFO Using 8 client processes
 0:01.94 INFO Starting http server on 127.0.0.1:8001
 0:01.94 INFO Starting http server on 127.0.0.1:8000
 0:01.95 INFO Starting https server on 127.0.0.1:8443
 0:02.41 SUITE_START: web-platform-test - running 1 tests
 0:02.43 TEST_START: /html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html
 0:02.43 WARNING Got command init_succeeded in state running
 0:02.72 pid:14527 Full command: /Users/twilco/projects/servo/target/debug/servo --hard-fail -u Servo/wptrunner -Z replace-surrogates -z http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/parse-a-sizes-attribute-standards-mode.html --user-stylesheet /Users/twilco/projects/servo/resources/ahem.css --certificate-path /Users/twilco/projects/servo/tests/wpt/web-platform-tests/tools/certs/cacert.pem
pid:14527 
 0:02.72 pid:14527 
 0:02.72 pid:14527 
 0:05.81 pid:14527 [components/script/dom/htmlimageelement.rs:395] &self.pending_request.borrow().is_none() = false
 0:05.81 pid:14527 [components/script/dom/htmlimageelement.rs:395] &self.pending_request.borrow().is_none() = false
 0:05.81 pid:14527 [components/script/dom/htmlimageelement.rs:395] &self.pending_request.borrow().is_none() = false
...lots more false dbg statements...
 0:06.86 TEST_END: Test OK. Subtests passed 0/169. Unexpected 154
FAIL <img srcset="/images/green-1x1.png?a2 300w, /images/green-16x16.png?a2 301w"> ref sizes="100vw" (standards mode) - assert_unreached: ref currentSrc was "" Reached unreachable code
check/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:13:9
Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
test@http://web-platform.test:8000/resources/testharness.js:544:30
check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:11:5
onload/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:26:5
onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:25:14
FAIL <img srcset="/images/green-1x1.png?b2 450w, /images/green-16x16.png?b2 451w"> ref sizes="100vw" (standards mode) - assert_unreached: ref currentSrc was "" Reached unreachable code
check/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:13:9
Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1908:25
test@http://web-platform.test:8000/resources/testharness.js:544:30
check@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:11:5
onload/<@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:26:5
onload@http://web-platform.test:8000/html/semantics/embedded-content/the-img-element/sizes/support/parse-a-sizes-attribute.js:25:14
...lots more failed tests...
 0:06.90 INFO Closing logging queue
 0:06.90 INFO queue closed
```

Perhaps I'm missing something obvious — any idea what is going on here?

jdm replied:
Wow. I can reproduce this behaviour change, and I'm really confused by it!

jdm replied:
Update: I've been digging into this. The test is triggering a situation where the img elements start a load, and then a window resize event is received (which is actually intended to be the window's initial size) which kicks off another image load while the first one is in progress. I still have no insight into why the behaviour changes how we write upgrade_pending_to_current.

jdm replied:
https://github.com/servo/servo/pull/24677 fixes the issues I described in my last comment and makes those `assert_unreached: ref currentSrc was "" Reached unreachable code` test failures go away.

twilco replied:
That's great — thanks for your continued investigation on this issue!  I'll keep an eye on that PR and rebase after it's landed.

jdm replied:
So I have a suspicion (based on https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3c8dee3b370e74a01e476e6f9b1152d9) that the reason the behaviour changed is that the original mem::swap code does not actually swap the values. The code that I suggested performs the swap as intended, and this exposed the problems that #24677 fixes.

jdm replied:
Yep - the presence of the `&mut` before each argument in the mem::swap means that we're swapping `&mut &mut` values. Yikes!

asajeffrey replied:
I suspect that the 
```rust
    mem::swap(
        &mut self.current_request.borrow_mut().deref_mut(),
        &mut pending_opt.get_or_insert(ImageRequest::new()),
    );
```
was meant to be
```rust
    mem::swap(
        &mut *self.current_request.borrow_mut().deref_mut(),
        &mut *pending_opt.get_or_insert(ImageRequest::new()),
    );
```

twilco replied:
Well that makes a lot more sense, good catch 😱

twilco committed:
Simplify nullification of pending request, fix step notation

twilco committed:
Don't use map — we need to return from function, not map closure

twilco committed:
Simplify upgrade_pending_to_current, .expect() pending request to be there

twilco commented:
Pushed some updates I have based on code review — will squash commits later.  Everything up to https://github.com/servo/servo/pull/24462/commits/61630f64fa6457e238fc440bcc13f82f59d58413 succeeds minus the 3 `.complete` failures.  Curious to try the WPT tests in CI to see if I can reproduce the `upgrade_pending_to_current` weirdness.

@bors-servo try=wpt

bors-servo commented:
@twilco: :key: Insufficient privileges: not in try users

bors-servo commented:
@twilco: :key: Insufficient privileges: not in try users

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 61630f64fa6457e238fc440bcc13f82f59d58413 with merge 652bc7499c2cd8de3d6fdfe89c319898d13029b2...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/13973)

twilco commented:
Thanks for kicking that off, @jdm.

Looks like the errors I wrote about in my novel above were placed into the `intermittents.log` category, while `filtered-wpt-errorsummary.log` has some `.complete` errors (as expected).  Not sure what protocol is — should I ignore the intermittent errors and focus on resolving the remaining `.complete` errors?

jdm commented:
@bors-servo try=wpt

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Trying commit 61630f64fa6457e238fc440bcc13f82f59d58413 with merge 54a4a26fbef8b3de8f1a7ef0e6ed42c1a5ff9a00...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/14056)

twilco committed:
Init pending_request to Unavailable when beginning image update

Done in an attempt to get .complete tests to pass

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 76d5fd84556e11d7d88377ce9e46697971d8b93a with merge 5a38178cf9cbaec7d190f71db393c0d707aff37f...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/14122)

jdm commented on components/script/dom/htmlimageelement.rs:
```diff
@@ -866,12 +921,10 @@ impl HTMLImageElement {
         {
-            let mut current_request = self.current_request.borrow_mut();
-            current_request.state = State::Unavailable;
+            if let Some(ref mut pending_request) = *self.pending_request.borrow_mut() {
```
I think we actually need the opposite logic:
```rust
let pending_request = self.pending_request.borrow_mut();
if pending_request.is_none() {
    *pending_request = Some(ImageRequest::new());
    pending_request.state = State::Unvailable;
}
```
This means we won't overwrite any already state that may already exist here, but if there is only a complete image then we'll add a new request that will trigger the conditional in HTMLImageElement::Complete correctly.

twilco replied:
Ok, that makes sense.  I just pushed a commit with the following:

```rust
let mut pending_opt = self.pending_request.borrow_mut();
if pending_opt.is_none() {
    *pending_opt = Some(ImageRequest::new());
}
```

WPT tests with back-to-back `<img>.complete` calls failed when I tried this locally on the first `<img>.complete` of each test, but perhaps CI WPT will prove that this change is fine.

This is different than your suggestion in that it doesn't explicitly set state to `Unavailable`, but `ImageRequest::new()` already sets the newly created image request state to `Unavailable`, so it should be effectively the same.  I tried to construct a new pending request and then explicitly set its state to `Unavailable`, but got [this error](https://stackoverflow.com/questions/58443400/rust-type-must-be-rooted).  If explicitly setting this state is desirable, I can continue my finaggling.

twilco committed:
Reset pending_request to unavailable only if it's None

This should help ensure <img>.complete responds appropriately in the face of back-to-back loads.

jdm commented:
@bors-servo try=wpt

bors-servo commented:
:hourglass: Trying commit 3b49970a2672eaf296dddfdb337f3202c6f4ee7d with merge 74f4bda96cfb3d2c353fa4836cfc58c10e1c4b50...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/14127)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #25515) made this pull request unmergeable. Please resolve the merge conflicts.
